src/Simulation/DepletionVoltage.jl

```diff
@@ -0,0 +1,164 @@
+    @assert first(potential_range) * last(potential_range) >= 0 "Please search for depletion voltage either in a fully positive or a fully negative range. Currently you were looking in the range
+        from $(first(potential_range)) to $(last(potential_range))"
+    potential_range = first(potential_range) >= 0 ? potential_range : reverse(potential_range)
+    
+    if verbose
+        @info "Looking for the depletion voltage applied to contact $(contact_id) "*
+              "in the range $(extrema(potential_range).*u"V") in steps of $(step(potential_range)*u"V")."
+    end
+    
+    for c in sim.detector.contacts
+        if c.potential == 0 && c.id != contact_id continue end
+        if ismissing(sim.weighting_potentials[c.id]) || sim.weighting_potentials[c.id].grid != sim.electric_potential.grid
+            @info "The weighting potential $(c.id) need to be defined on the same grid as the electric potential. Adjusting weighting potential $(c.id) now."
+            _adapt_weighting_potential_to_electric_potential_grid!(sim, c.id)
+        end
+    end
+    
+    # ϕρ is the electric potential resulting only from the impurity density
+    # and setting all potentials on the contacts to zero
+    ϕρ = deepcopy(sim.electric_potential.data)
+    for c in sim.detector.contacts
+        if c.potential == 0 continue end
+        ϕρ .-= c.potential * sim.weighting_potentials[c.id].data
+    end
+    
+    inside = findall(p -> p & bulk_bit > 0 && p & update_bit > 0, sim.point_types.data)
+    
+    ϕmin, ϕmax = extrema((ϕρ .+ potential_range[1] * sim.weighting_potentials[contact_id].data)[inside])
+    initial_depletion::Bool = ϕmax - ϕmin < abs(potential_range[1])
+    depletion_voltage::T = NaN
+    start_local_search::T = 0
+    
+    @showprogress for U in potential_range
+        ϕmin, ϕmax = extrema((ϕρ .+ T(U) * sim.weighting_potentials[contact_id].data)[inside])
+        depleted = ϕmax - ϕmin < abs(U)
+        if (initial_depletion && !depleted) || (!initial_depletion && depleted)
+            if T(U)>=0
+                start_local_search = T(U)
+            else
+                start_local_search = T(U)
+            end
+            break
+        end
+    end
+    local_range::AbstractRange = range(start_local_search, last(potential_range), step = step(potential_range))
+    size_data = size(sim.point_types.data)
+    
+    if size_data[2] == 1
+        Ix = CartesianIndex(1,0,0)
+        Iz = CartesianIndex(0,0,1)
+        neighbours = [Ix,-Ix,Iz,-Iz]
+    else
+        Ix = CartesianIndex(1,0,0)
+        Iy = CartesianIndex(0,1,0)
+        Iz = CartesianIndex(0,0,1)
+        neighbours = [Ix,-Ix,Iy,-Iy,Iz,-Iz]
+    end
+
+    scale = start_local_search
+    undepleted = 0
+    for (i,idx) in enumerate(inside)
+        if !(0 in [size_data[j] == 1 || idx[j] != 1 && idx[j] != size_data[j] for j in 1:length(idx)])#idx[1]!=1 && idx[1]!=size_data[1] && idx[2]!=1 && idx[2]!=size_data[2] && idx[3]!=1 && idx[3]!=size_data[3]
```

Comment:
Are you sure that this statement is executed correctly?
You might want to add brackets around statements to ensure that `&&` and `||` are not mixed up.

Is it:
- `(size_data[j] == 1) || (idx[j] != 1 && idx[j] != size_data[j])` or
- `(size_data[j] == 1 || idx[j] != 1) && (idx[j] != size_data[j])` ?

Reply:
I checked, it does the right thing (if `size_data[j]` only has one entry of if we are at the beginning or end)